### PR TITLE
[StandardInstrumentations] Support -print-after-pass-number option

### PR DESCRIPTION
**Pull request by thevinster**
There's already support for `-print-before-pass-number`, so it makes sense that we also have a `-print-after-pass-number`. This is especially useful if you want to print the IR after the very last pass without resorting to `-print-after-all` and combing through stderr or the IR file directory.

### CODE REVIEW

File: llvm/lib/Passes/StandardInstrumentations.cpp [hunk 1]
```diff
@@ -802,6 +807,8 @@ void PrintIRInstrumentation::printBeforePass(StringRef PassID, Any IR) {
       (shouldPrintBeforePass(PassID) || shouldPrintAfterPass(PassID)))
     DumpIRFilename = fetchDumpFilename(PassID, IR);
 
+  ++CurrentPassNumber;
```
Comment (aeubanks) [inline llvm/lib/Passes/StandardInstrumentations.cpp:810]: this will make pass numbers non-consecutive when we have passes we don't print IR for. is this change necessary? can you find a way to preserve the old behavior? (we should probably add a test that the pass numbers are consecutive)

Comment (thevinster) [reply llvm/lib/Passes/StandardInstrumentations.cpp:810]: print-at-pass-number.ll does test the pass numbers are consecutive (although it doesn't test passes where the IR isn't printed). If we want to preserve the existing behavior, I can hoist the `shouldPrintIR` call above. The reason for hoisting incrementing `CurrentPassNumber` is so to have the `shouldPrintAfterPass` run after which is needed in the `printAfterPass` call.

Comment (thevinster) [reply llvm/lib/Passes/StandardInstrumentations.cpp]: Moved the `CurrentPassNumber` back to where it was to preserve the behavior of consecutive pass numbers for printing filtered IR. I also added a test for it in the same file. The new solution is to push a new PassRunDescriptor after `CurrentPassNumber` has been incremented.

Comment (thevinster) [reply llvm/lib/Passes/StandardInstrumentations.cpp]: Ugh nevermind, this approach doesn't work as it doesn't clean the PassRunDescriptor stack up. The new approach is to implement a separate check that only pushes the PassRunDescriptor when the you are running with `-print-after-pass-number` option.

**Comment by thevinster** [comment]
Ping. @aeubanks I think this version should satisfy your requirements.

File: llvm/lib/Passes/StandardInstrumentations.cpp [hunk 2]
```diff
@@ -929,6 +942,11 @@ bool PrintIRInstrumentation::shouldPrintAfterPass(StringRef PassID) {
   return is_contained(printAfterPasses(), PassName);
 }
 
+bool PrintIRInstrumentation::shouldPrintAfterPassNumberCheck() {
```
Comment (aeubanks) [inline llvm/lib/Passes/StandardInstrumentations.cpp:945]: `shouldPrintAfterPassNumberCheck`/`shouldPrintAfterPassNumber` is a bit confusing. perhaps `shouldPrintAfter/BeforePassNumber` should be `shouldPrintAfter/BeforeSomePassNumber`, and `shouldPrintAfter/BeforePassNumberCheck` should be `shouldPrintAfter/BeforeCurrentPassNumber`?

**Review by aeubanks — APPROVED** [review]
